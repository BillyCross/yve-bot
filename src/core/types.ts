import fetch = require('fetch-everywhere');

import { uniq } from 'lodash';

import { DefineModule } from './module';
import { ValidatorError } from './exceptions';
import { Executors } from './executors';
import { YveBot } from './bot';
import * as utils from './utils';

import { Rule, Answer, RuleType } from '../types';


const executors = new Executors;


const types: { [name: string]: RuleType } = {
  Any: { executors: [{}] },

  String: {
    executors: [{
      transform: async (value: Answer) => String(value),
      validators: [
        {
          string: true,
          warning: 'Invalid string',
        },
      ],
    }],
  },

  Number: {
    executors: [{
      transform: async (value: Answer) => Number(value),
      validators: [
        {
          number: true,
          warning: 'Invalid number',
        },
      ],
    }],
  },

  SingleChoice: {
    executors: [{
      transform: async (value: Answer | Answer[], rule: Rule) => {
        const option = utils.findOptionByAnswer(rule.options, value);
        if (!option) {
          return undefined;
        }
        return option.value || option.label;
      },
      validators: [
        {
          function: (answer: Answer | Answer[], rule: Rule) =>
            !!utils.findOptionByAnswer(rule.options, answer),
          warning: 'Unknown option',
        },
      ],
    }],
  },

  MultipleChoice: {
    executors: [{
      transform: async (answer: Answer | Answer[], rule: Rule) => {
        let values;
        if (answer instanceof Array) {
          values = answer;
        } else {
          const options = rule.options.map(o => String(o.value || o.label));
          values = utils.identifyAnswersInString(String(answer), options);
        }
        return uniq(values
          .map(value => {
            const option = utils.findOptionByAnswer(rule.options, value);
            if (!option) {
              return undefined;
            }
            return option.value || option.label;
          })
          .filter(x => x));
      },
      validators: [
        {
          function: (answer: Answer | Answer[], rule: Rule) => {
            const answers = utils.ensureArray(answer);
            const options = rule.options.map(o => String(o.value || o.label));
            return answers.every(a => options.some(o => utils.isMatchAnswer(a, o)));
          },
          warning: 'Unknown options',
        },
      ],
    }],
  },

  StringSearch: {
    executors: [
      {}, // necessary to read user's input and apply Rule's validators to it
      {
        transform: async (answer: Answer, rule: Rule, bot: YveBot) => {
          /*
           * Server MUST return a JSON with a list of objects following the format:
           * [
           *   { label: String, value: Any }
           * ]
           * Or you will need to set rule.config.translate object with your
           * server response's objects properties to this expected format
           * - translate:
           *   - label: myProperty1
           *   - value: myProperty2
           */
          const { apiURI, apiQueryParam, translate } = rule.config;
          const searchURI = `${apiURI}?${apiQueryParam}=${encodeURIComponent(String(answer))}`;

          bot.dispatch('typing');
          return fetch(searchURI)
            .then(res => res.json())
            .then(list => {
              if (!translate) {
                return list;
              }
              const { label, value } = translate;
              return list.map(obj => ({ label: obj[label], value: obj[value] }));
            });
        },
        validators: [{
          function: (results: any, rule: Rule) => {
            if (results.length === 0) {
              throw new ValidatorError(rule.config.messages.noResults, rule);
            }
            return true;
          }
        }]
      },
      {
        transform: async (results: any, rule: Rule, bot: YveBot) => {
          const { messages } = rule.config;
          let options, message;
          if (results.length === 1) {
            message = `${messages.didYouMean}: ${results[0].label}?`;
            options = [
              { label: 'Sim', value: results[0].value },
              { label: 'NÃ£o', value: null },
            ];
          } else {
            message = `${messages.multipleResults}:`;
            options = results.concat([{
              label: messages.noneOfAbove,
              value: null,
            }]);
          }

          bot.talk(message, { type: 'SingleChoice', options });
        },
        validators: [...executors.WaitForUserInput.validators]
      },
      {
        validators: [{
          function: (answer: any, rule: Rule, bot: YveBot) => {
            const { messages } = rule.config;
            if (!answer) {
              bot.store.unset(`executors.${rule.name}.currentIdx`);
              bot.talk(messages.wrongResult);
              return (
                executors.WaitForUserInput.validators
                  .reduce((list, validator) => list.concat(
                    Object.keys(validator).map(k => validator[k])
                  ), [])
                  .filter(validator => validator instanceof Function)
                  .every(validator => validator('', rule))
              );
            }
            return true;
          }
        }]
      }
    ],
  }
};

export class Types extends DefineModule {
  public Any: RuleType;
  public String: RuleType;
  public Number: RuleType;
  public SingleChoice: RuleType;
  public MultipleChoice: RuleType;
  public StringSearch: RuleType;

  constructor() {
    super();
    this.define(types);
  }
}
