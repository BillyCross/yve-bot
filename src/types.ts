import { YveBot } from './core/bot';

export interface YveBotOptions {
  timePerChar?: number;
  enableWaitForSleep?: boolean;
  rule?: Rule;
  context?: Context;
}

export interface Flow {
  flow: string;
  rules: Rule[];
}

export interface Rule {
  flow?: string;
  name?: string;
  type?: string;
  output?: string;
  message?: string;
  delay?: number;
  sleep?: number;
  actions?: RuleAction[];
  preActions?: RuleAction[];
  replyMessage?: string;
  options?: RuleOption[];
  validators?: RuleValidator[];
  config?: { [name: string]: any };
  next?: RuleNext;
  exit?: boolean;
}

export interface RuleOption {
  label?: string;
  value?: string | number;
  synonyms?: string[];
  next?: RuleNext;
}

export interface RuleAction {
  [name: string]: any;
}

export interface RuleValidator {
  [name: string]: any;
}

export interface RuleType {
  executors?: RuleTypeExecutor[];
}

export interface RuleTypeExecutor {
  validators?: RuleValidator[];
  transform?: RuleTypeTransform;
}

export type RuleTypeTransform = (
  (value: any, rule?: Rule, bot?: YveBot) => Promise<any>
);

export type RuleNext = string;

export type Context = {
  [key: string]: any;
};

export type Indexes = {
  [ruleName: string]: number;
};

export type Answer = string | number;

export type EventName = 'start' | 'end' | 'error' | 'receive' | 'hear' | 'talk' | 'typed' | 'typing' | 'storeChanged';

export interface ChatOptions {
  target?: string;
  name?: string;
  inputPlaceholder?: string;
  inputPlaceholderSingleChoice?: string;
  inputPlaceholderMutipleChoice?: string;
  doneMultipleChoiceLabel?: string;
  andSeparatorText?: string;
  submitLabel?: string;
  timestampable?: boolean;
  timestampFormatter?: (ts: number) => string;
  autoFocus?: boolean;
  yveBotOptions?: YveBotOptions;
}

export type ChatMessageSource = 'BOT' | 'USER';
