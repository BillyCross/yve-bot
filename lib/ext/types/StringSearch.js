!function(e,n){"object"==typeof exports&&"undefined"!=typeof module?module.exports=n(require("isomorphic-unfetch"),require("../../core")):"function"==typeof define&&define.amd?define(["isomorphic-unfetch","../../core"],n):e.YveBotTypesStringSearch=n(null,e.YveBot)}(this,function(e,n){"use strict";function t(e,n,t,r){return new(t||(t=Promise))(function(o,i){function u(e){try{c(r.next(e))}catch(e){i(e)}}function a(e){try{c(r.throw(e))}catch(e){i(e)}}function c(e){e.done?o(e.value):new t(function(n){n(e.value)}).then(u,a)}c((r=r.apply(e,n||[])).next())})}function r(e,n){function t(t){return function(u){return function(t){if(r)throw new TypeError("Generator is already executing.");for(;a;)try{if(r=1,o&&(i=o[2&t[0]?"return":t[0]?"throw":"next"])&&!(i=i.call(o,t[1])).done)return i;switch(o=0,i&&(t=[0,i.value]),t[0]){case 0:case 1:i=t;break;case 4:return a.label++,{value:t[1],done:!1};case 5:a.label++,o=t[1],t=[0];continue;case 7:t=a.ops.pop(),a.trys.pop();continue;default:if(i=a.trys,!(i=i.length>0&&i[i.length-1])&&(6===t[0]||2===t[0])){a=0;continue}if(3===t[0]&&(!i||t[1]>i[0]&&t[1]<i[3])){a.label=t[1];break}if(6===t[0]&&a.label<i[1]){a.label=i[1],i=t;break}if(i&&a.label<i[2]){a.label=i[2],a.ops.push(t);break}i[2]&&a.ops.pop(),a.trys.pop();continue}t=n.call(e,a)}catch(e){t=[6,e],o=0}finally{r=i=0}if(5&t[0])throw t[1];return{value:t[0]?t[1]:void 0,done:!0}}([t,u])}}var r,o,i,u,a={label:0,sent:function(){if(1&i[0])throw i[1];return i[1]},trys:[],ops:[]};return u={next:t(0),throw:t(1),return:t(2)},"function"==typeof Symbol&&(u[Symbol.iterator]=function(){return this}),u}return(n=n&&n.hasOwnProperty("default")?n.default:n).types.defineExtension("StringSearch",{executors:[{},{transform:function(e,n,o){return t(void 0,void 0,void 0,function(){var t,i,u,a,c,l;return r(this,function(r){return t=n.config,i=t.apiURI,u=t.apiQueryParam,a=t.translate,c=t.messages,l=i+"?"+u+"="+encodeURIComponent(String(e)),o.dispatch("typing"),[2,fetch(l).then(function(e){return e.json()}).then(function(e){if(0===e.length)throw new o.exceptions.ValidatorError(c.noResults,n);return e}).then(function(e){if(!a)return e;var n=a.label,t=a.value;return e.map(function(e){return{label:e[n],value:e[t]}})})]})})}},{transform:function(e,n,o){return t(void 0,void 0,void 0,function(){var t,i,u;return r(this,function(r){if(!0===o.store.get("stringsearch."+n.name+".wait"))return o.store.unset("stringsearch."+n.name+".wait"),[2,e];throw t=n.config.messages,1===e.length?(u=t.didYouMean+": "+e[0].label+"?",i=[{label:t.yes,value:e[0].value},{label:t.no,value:null}]):(u=t.multipleResults+":",i=e.concat([{label:t.noneOfAbove,value:null}])),o.store.set("stringsearch."+n.name+".wait",!0),o.talk(u,{type:"SingleChoice",options:i}),new o.exceptions.PauseRuleTypeExecutors(n.name)})})}},{validators:[{function:function(e,n,t){var r=n.config.messages;if(!e)throw t.store.unset("executors."+n.name+".currentIdx"),t.talk(r.wrongResult),new t.exceptions.PauseRuleTypeExecutors(n.name);return!0}}]}]})});
