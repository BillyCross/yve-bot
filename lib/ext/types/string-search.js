(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('isomorphic-unfetch'), require('../../core')) :
	typeof define === 'function' && define.amd ? define(['isomorphic-unfetch', '../../core'], factory) :
	(factory(null,global.YveBot));
}(this, (function (isomorphicUnfetch,YveBot) { 'use strict';

YveBot = YveBot && YveBot.hasOwnProperty('default') ? YveBot['default'] : YveBot;

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */













function __awaiter(thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var _this = undefined;
var _a = YveBot.exceptions;
var PauseRuleTypeExecutors = _a.PauseRuleTypeExecutors;
var ValidatorError = _a.ValidatorError;
YveBot.types.define('StringSearch', {
    executors: [
        {},
        {
            transform: function (answer, rule, bot) { return __awaiter(_this, void 0, void 0, function () {
                var _a, apiURI, apiQueryParam, translate, messages, searchURI;
                return __generator(this, function (_b) {
                    _a = rule.config, apiURI = _a.apiURI, apiQueryParam = _a.apiQueryParam, translate = _a.translate, messages = _a.messages;
                    searchURI = apiURI + "?" + apiQueryParam + "=" + encodeURIComponent(String(answer));
                    bot.dispatch('typing');
                    return [2, fetch(searchURI)
                            .then(function (res) { return res.json(); })
                            .then(function (list) {
                            if (list.length === 0) {
                                throw new ValidatorError(messages.noResults, rule);
                            }
                            return list;
                        })
                            .then(function (list) {
                            if (!translate) {
                                return list;
                            }
                            var label = translate.label, value = translate.value;
                            return list.map(function (obj) { return ({ label: obj[label], value: obj[value] }); });
                        })];
                });
            }); },
        },
        {
            transform: function (results, rule, bot) { return __awaiter(_this, void 0, void 0, function () {
                var messages, options, message;
                return __generator(this, function (_a) {
                    messages = rule.config.messages;
                    if (results.length === 1) {
                        message = messages.didYouMean + ": " + results[0].label + "?";
                        options = [
                            { label: messages.yes, value: results[0].value },
                            { label: messages.no, value: null },
                        ];
                    }
                    else {
                        message = messages.multipleResults + ":";
                        options = results.concat([{
                                label: messages.noneOfAbove,
                                value: null,
                            }]);
                    }
                    bot.talk(message, { type: 'SingleChoice', options: options });
                    throw new PauseRuleTypeExecutors(rule.name);
                });
            }); },
        },
        {
            validators: [{
                    function: function (answer, rule, bot) {
                        var messages = rule.config.messages;
                        if (!answer) {
                            bot.store.unset("executors." + rule.name + ".currentIdx");
                            bot.talk(messages.wrongResult);
                            throw new PauseRuleTypeExecutors(rule.name);
                        }
                        return true;
                    },
                }],
        },
    ],
});

})));
