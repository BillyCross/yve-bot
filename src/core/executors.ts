import { Answer, IRule, IRuleTypeExecutor } from '../types';
import { PauseRuleTypeExecutors } from './exceptions';
import { DefineModule } from './module';

const executors: { [name: string]: IRuleTypeExecutor } = {
  WaitForUserInput: {
    validators: [{
      function: (_: Answer, rule: IRule): void => {
        throw new PauseRuleTypeExecutors(rule.name);
      },
    }],
  },
};

export class Executors extends DefineModule {
  public WaitForUserInput: IRuleTypeExecutor;

  constructor() {
    super('executors');
    this.define(executors);
  }
}
