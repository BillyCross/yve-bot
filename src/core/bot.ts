import { YveBotOptions, Rule, Flow, Answer, Context, EventName } from '../types';
import { Store, StoreData } from './store';
import { Controller } from './controller';
import { Actions } from './actions';
import { Types } from './types';
import { Executors } from './executors';
import { Validators } from './validators';

import * as Exceptions from './exceptions';

function sanitizeRule(rule: Rule): Rule {
  if (typeof rule === 'string') {
    return { message: rule };
  }
  if (['SingleChoice', 'MultipleChoice'].indexOf(rule.type) >= 0) {
    rule.options = (rule.options || []).map(o => {
      if (typeof o === 'string') {
        return { value: o };
      }
      if (typeof o.synonyms === 'string' && !!o.synonyms) {
        const synonyms: string = o.synonyms;
        o.synonyms = synonyms.split(',').map(s => s.trim());
      }
      return o;
    });
  }
  return rule;
}

function convertToRules(inputs: (Flow|Rule)[]): Rule[] {
  let rules: Rule[] = [];
  inputs.forEach(input => {
    const isFlow = typeof input !== 'string' && 'rules' in input && 'flow' in input;
    if (isFlow) {
      const flow = input as Flow;
      rules = rules.concat(
        flow.rules.map(
          rule => Object.assign({}, sanitizeRule(rule), {
            flow: input.flow,
          })
        )
      );
    } else {
      rules.push(sanitizeRule(input as Rule));
    }
  });
  return rules;
}

export class YveBot {
  static types: Types;
  static actions: Actions;
  static executors: Executors;
  static validators: Validators;
  static exceptions: any;

  private handlers: { [handler: string]: Array<() => any> };
  private _rules?: Rule[];

  public options: YveBotOptions;
  public rules: Rule[];
  public controller: Controller;
  public store: Store;
  public sessionId: string;

  constructor(rules: (Rule|Flow)[], customOpts?: YveBotOptions) {
    const DEFAULT_OPTS: YveBotOptions = {
      enableWaitForSleep: true,
      timePerChar: 40,
    };

    this.sessionId = 'session';
    this.options = Object.assign({}, DEFAULT_OPTS, customOpts);
    this.rules = convertToRules(rules);
    this.handlers = {};

    this.store = new Store(this);
    this.controller = new Controller(this);

    if (this.options.context) {
      this.store.set('context', this.options.context);
    }

    this.on('error', err => { throw err; });
  }

  public get context(): Context {
    return this.store.get('context');
  }

  public get types() { return YveBot.types; }
  public get actions() { return YveBot.actions; }
  public get executors() { return YveBot.executors; }
  public get validators() { return YveBot.validators; }

  on(evt: EventName, fn: (...args: any[]) => any): this {
    const isUniqueType = ['error'].indexOf(evt) >= 0;
    if (!isUniqueType && evt in this.handlers) {
      this.handlers[evt].push(fn);
    } else {
      this.handlers[evt] = [fn];
    }
    return this;
  }

  start(): this {
    this.dispatch('start');
    this.controller.run().catch(this.tryCatch.bind(this));
    return this;
  }

  end(): this {
    this.dispatch('end', this.store.output());
    return this;
  }

  talk(message: string, opts?: object): this {
    const rule = Object.assign({}, this.options.rule, opts || {});
    this.controller.sendMessage(message, rule);
    return this;
  }

  hear(answer: Answer | Answer[]): this {
    this.controller.receiveMessage(answer).catch(this.tryCatch.bind(this));
    return this;
  }

  dispatch(name: EventName, ...args) {
    if (name in this.handlers) {
      this.handlers[name].forEach(fn => fn(...args, this.sessionId));
    }
  }

  session(
    id: string,
    opts: { context?: Context, store?: StoreData, rules?: Rule[] } = {},
  ): this {
    this.sessionId = id;

    if (opts.rules) {
      this._rules = this.rules;
      this.rules = opts.rules.map(rule => sanitizeRule(rule));
    } else {
      this.rules = this._rules || this.rules;
    }

    if (opts.store) {
      this.store.replace(opts.store);
    } else {
      this.store.reset();
      this.controller.reindex();
    }

    if (opts.context) {
      this.store.set('context', opts.context);
    }

    return this;
  }

  private tryCatch(err: Error) {
    this.dispatch('error', err);
    this.end();
  }
}

YveBot.types = new Types;
YveBot.actions = new Actions;
YveBot.executors = new Executors;
YveBot.validators = new Validators;
YveBot.exceptions = Exceptions;
