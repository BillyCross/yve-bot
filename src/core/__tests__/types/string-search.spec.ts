import { Types } from '../../types';
import { Executors } from '../../executors';
import { YveBot } from '../../bot';

import * as fetchMock from 'fetch-mock';

describe('StringSearch', () => {
  const { executors } = (new Types).StringSearch;

  test('properties', () => {
    expect(executors).toHaveLength(4);
    expect(executors[0]).toEqual({});
    executors.slice(1).map(exe => {
      expect(exe).toHaveProperty('validators');
    });
    executors.slice(1, -1).map(exe => {
      expect(exe).toHaveProperty('transform');
    });
  });

  describe('validators', () => {
    test('first executor', () => {
      const { validators } = executors[1];
      expect(validators).toHaveLength(1);

      const noResults = 'noResults msg first-executor-validator';
      const rule = {
        rand: Math.random(),
        config: { messages: { noResults } }
      };
      const { function: validator } = validators[0];
      const validate = (result) => validator(result, rule);

      expect(validate([1])).toBeTruthy();
      expect(() => validate([])).toThrowErrorMatchingSnapshot();
    });

    test('wait for input executor', () => {
      const { validators } = executors[2];
      expect(validators).toEqual((new Executors).WaitForUserInput.validators);
    });

    test('last executor', () => {
      const { validators } = executors[3];
      expect(validators).toHaveLength(1);

      const bot = { talk: jest.fn(), store: { unset: jest.fn() } };
      const wrongResult = 'wrongResult msg last-executor-validator';
      const rule = {
        name: 'testRule',
        rand: Math.random(),
        config: { messages: { wrongResult } }
      };

      const { function: validator } = validators[0];
      expect(validator('something', rule, bot)).toBeTruthy();
      expect(bot.talk).not.toHaveBeenCalled();
      expect(bot.store.unset).not.toHaveBeenCalled();

      expect(() => validator(null, rule, bot)).toThrowErrorMatchingSnapshot();
      expect(bot.talk).toHaveBeenCalledWith(wrongResult);
      expect(bot.store.unset).toHaveBeenCalledWith(`executors.${rule.name}.currentIdx`);
    });
  });

  describe('transforms', () => {
    afterEach(() => {
      fetchMock.restore();
    });

    test('call server', async () => {
      const customResponse = { rand: Math.random() };

      const { transform } = executors[1];
      const input = 'test input';
      const apiURI = 'https://myserver.test/search';
      const apiQueryParam = 'param';
      const rule = {
        rand: Math.random(),
        config: { apiURI, apiQueryParam },
      };
      const bot = new YveBot([], { rule, enableWaitForSleep: false });

      const expectedURI = `${apiURI}?${apiQueryParam}=${encodeURIComponent(input)}`;
      fetchMock.mock(expectedURI, {
        status: 200,
        body: [customResponse]
      });

      const result = await transform(input, rule, bot);
      expect(fetchMock.called(expectedURI)).toBeTruthy();
      expect(result).toEqual([customResponse]);
    });

    test('translates server response', async () => {
      const customResponse = { test1: 'fake label', test2: 'fake val' };

      const { transform } = executors[1];
      const input = 'test input';
      const apiURI = 'https://myserver.test/search';
      const apiQueryParam = 'param';
      const translate = { label: 'test1', value: 'test2' };
      const rule = {
        rand: Math.random(),
        config: { apiURI, apiQueryParam, translate },
      };
      const bot = new YveBot([], { rule, enableWaitForSleep: false });

      fetchMock.mock('*', {
        status: 200,
        body: [customResponse]
      });

      const result = await transform(input, rule, bot);
      expect(result).toEqual([customResponse].map(obj => ({
        value: obj.test2, label: obj.test1
      })));
    });

    test('organize server response with 1 result', async () => {
      const serverResponse = { label: 'fake1', value: 'fake2' };

      const { transform } = executors[2];
      const messages = { didYouMean: 'você quis dizer' };
      const rule = {
        rand: Math.random(),
        config: { messages },
      };
      const bot = new YveBot([], { rule, enableWaitForSleep: false });
      bot.talk = jest.fn();

      await transform([serverResponse], rule, bot);

      expect(bot.talk).toHaveBeenCalled();
      expect(bot.talk).toHaveBeenCalledWith(`${messages.didYouMean}: ${serverResponse.label}?`, {
        type: 'SingleChoice',
        options: [
          { label: 'Sim', value: serverResponse.value },
          { label: 'Não', value: null },
        ]
      });
    });

    test('organize server response with multiple results', async () => {
      const serverResponse = [
        { label: 'label1', value: 'value1' },
        { label: 'label2', value: 'value2' },
      ];

      const { transform } = executors[2];
      const messages = { multipleResults: 'multipleResults', noneOfAbove: 'noneOfAbove' };
      const rule = {
        rand: Math.random(),
        config: { messages },
      };
      const bot = new YveBot([], { rule, enableWaitForSleep: false });
      bot.talk = jest.fn();

      await transform(serverResponse, rule, bot);

      expect(bot.talk).toHaveBeenCalled();
      expect(bot.talk).toHaveBeenCalledWith(`${messages.multipleResults}:`, {
        type: 'SingleChoice',
        options: serverResponse.concat([{
          label: messages.noneOfAbove,
          value: null,
        }])
      });
    });
  });
});
