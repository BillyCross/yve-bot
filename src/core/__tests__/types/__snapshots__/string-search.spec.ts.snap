// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringSearch validators first executor 1`] = `"noResults msg first-executor-validator"`;

exports[`StringSearch validators last executor 1`] = `"Rule \\"testRule\\" must pause execution"`;
